db/silkworm/etl/file_provider.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+   Copyright 2020 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include <silkworm/etl/file_provider.hpp>
+
+namespace silkworm::etl{
+
+FileProvider::FileProvider(Buffer * buffer, int id) {
+    filename_ = "tmp" + std::to_string(id);
+    file_.open(filename_, std::ios_base::in | std::ios_base::out | std::ios_base::trunc);
+    auto entries{buffer->get_entries()};
+    for(auto entry: entries) {
```

Comment:
I'd avoid making complex operations in class constructor as it gets difficult to determine if the initialization fails or any of the operations fail.
In this case however your making multiple copies of the same data:
- you're making a copy of each entry in entries
- you're copying data into a string
- you're multiple casting string into bytes

Also I don't understand the point of saving data lengths in BE. Native endianness should be enough as file contents are supposedly read from within the same architecture ... this saves the overhead of byte swapping.
A suggestion follows

Please also take note that this code does not account possible errors while writing files (eg. disk full)

```suggestion
    filename_ = "tmp" + std::to_string(id);
    file_.open(filename_, std::ios_base::in | std::ios_base::out | std::ios_base::trunc);

    union head_t {
        uint32_t lengths[2];
        uint8_t bytes[8];
    };

    head_t head{};

    for(const auto& entry: buffer->get_entries()) {
        head.lengths[0] = (uint32_t)entry.key.size();
        head.lengths[1] = (uint32_t)entry.value.size();
        file_.write((const char *)head.bytes, 8);
        file_.write((const char *)entry.key.data(), entry.key.size());
        file_.write((const char *)entry.value.data(), entry.value.size());
    }
    file_.seekp(0);
```